Nix Flake can't download Kubeseal deps

### Describe the bug

The Nix Flake fails to build the Kubeseal package. It looks like the problem is that the Go build can't download some of the Kubeseal deps from proxy.golang.org.


### To Reproduce

1. Create a multipass VM: `multipass launch --name test --cpus 2 --mem 4G --disk 40G 20.04`
2. Shell into it: `multipass shell test`
3. Install Nix: `sh <(wget -qO- https://nixos.org/nix/install)`
4. Set up Nix env: `. /home/ubuntu/.nix-profile/etc/profile.d/nix.sh`
5. Create Nix config dir: `mkdir -p ~/.config/nix`
6. Enable Nix Flakes: `echo 'experimental-features = nix-command flakes' >> ~/.config/nix/nix.conf`
7. Clone KITT4SME live repo: `git clone https://github.com/c0c0n3/kitt4sme.live`
8. Enter KITT4SME's Nix shell: `cd kitt4sme.live/nix && nix shell`

You should see an error similar to the one below:

```console
error: builder for '/nix/store/4vzgbmn0mqqadi7swf278qdj97y26ysn-kubeseal-0.17.5.drv' failed with exit code 1;
       last 10 log lines:
       > pkg/apis/sealed-secrets/v1alpha1/sealedsecret_expansion.go:14:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:47564->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:28:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:47564->[::1]:53: read: connection refused
       > pkg/apis/sealed-secrets/v1alpha1/register.go:7:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:47564->[::1]:53: read: connection refused
       > pkg/apis/sealed-secrets/v1alpha1/register.go:8:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51224->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:30:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51224->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:43:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51224->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:31:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51224->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:32:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51224->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:33:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51224->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:34:2: k8s.io/klog/v2@v2.40.1: Get "https://proxy.golang.org/k8s.io/klog/v2/@v/v2.40.1.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:55715->[::1]:53: read: connection refused
       For full logs, run 'nix log /nix/store/4vzgbmn0mqqadi7swf278qdj97y26ysn-kubeseal-0.17.5.drv'.
error: 1 dependencies of derivation '/nix/store/gcxa0icdrv6d5zy9qh84zlyrx8hi1vz9-kitt4sme-cluster-shell.drv' failed to build
```


### Expected behavior

Nix should build the Kubeseal package and you should be able to enter the KITT4SME's shell.

### Additional context

See
- https://github.com/c0c0n3/kitt4sme.live/discussions/166
- https://github.com/c0c0n3/kitt4sme.live/discussions/144#discussioncomment-2804362

So I've done some extensive testing and debugging, but the reason why we're getting that "connection refused" error is still shrouded in mystery. For the record, here's what to do to reproduce the error in an isolated environment with just the kubeseal package.

### VM

Get Multipass to spin up a 20.4 Ubuntu VM and shell into it. (Architecture: x86_64)

```bash
$ multipass launch --name debug --cpus 2 --mem 4G --disk 40G 20.04
$ multipass shell debug
```

The rest of the steps below assume you're running the commands inside this shell on the Ubuntu VM.


### Nix

Install Nix on the VM and enable Flakes.

```bash
$ sh <(wget -qO- https://nixos.org/nix/install)
$ . /home/ubuntu/.nix-profile/etc/profile.d/nix.sh
$ mkdir -p ~/.config/nix
$ echo 'experimental-features = nix-command flakes' >> ~/.config/nix/nix.conf
```


### Test repo

Create a git repo with the `flake.nix` and `kubeseal.nix` files below in it.

```bash
$ mkdir repo && cd repo
$ git init
```

`flake.nix`

```nix
{
  description = "Debug me!";

  inputs = {
    nixpkgs.url = "github:NixOs/nixpkgs/nixos-21.11";
  };

  outputs = { self, nixpkgs }:
    let
      system = "x86_64-linux";
      pkgs = nixpkgs.legacyPackages.${system};
      kubeseal = pkgs.callPackage ./kubeseal.nix {};
    in {
      packages.${system}.kubeseal = kubeseal;
      defaultPackage.${system} = pkgs.buildEnv {
        name = "shell";
        paths = [ kubeseal ];
      };
    };
}
```

`kubeseal.nix`

```nix
#
# Custom Sealed Secrets package.
#
# So I've rolled out a bare build recipe whipped together by looking
# at what the kubeseal Makefile target does. To be replaced with the
# official Nixpkgs expression.
#
{ lib, stdenv, go_1_17, fetchFromGitHub }:

stdenv.mkDerivation rec {
  pname = "kubeseal";
  version = "0.17.5";

  src = fetchFromGitHub {
    owner = "bitnami-labs";
    repo = "sealed-secrets";
    rev = "v${version}";
    sha256 = "sha256-cqOSMAagefKQiKKtgVbk1UFKYGBXQleJ1pgcJ/VyOnM=";
  };

  GO_LD_FLAGS = "-s -w -X main.VERSION=${version}";

  buildInputs = [ go_1_17 ];

  buildPhase = ''
    mkdir -p $out/mod-cache
    export GOMODCACHE=$out/mod-cache

    mkdir -p $out/build-cache
    export GOCACHE=$out/build-cache

    go build -o kubeseal -ldflags "$GO_LD_FLAGS" ./cmd/kubeseal
  '';
  installPhase = ''
    mkdir -p $out/bin
    cp kubeseal $out/bin
  '';

}
```

Now add the two files to the git index.

```bash
$ git add .
```


### Flake build

The moment we've been eagerly waiting for. Drum roll...

```bash
$ nix build
```

It should bomb out with an error similar to this:

```
error: builder for '/nix/store/h0bnm7dvj7jg6vhg86a9lalyqnz687wb-kubeseal-0.17.5.drv' failed with exit code 1;
       last 10 log lines:
       > pkg/apis/sealed-secrets/v1alpha1/sealedsecret_expansion.go:14:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51547->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:28:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51547->[::1]:53: read: connection refused
       > pkg/apis/sealed-secrets/v1alpha1/register.go:7:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:51547->[::1]:53: read: connection refused
       > pkg/apis/sealed-secrets/v1alpha1/register.go:8:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:58276->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:30:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:58276->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:43:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:58276->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:31:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:58276->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:32:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:58276->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:33:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:58276->[::1]:53: read: connection refused
       > cmd/kubeseal/main.go:34:2: k8s.io/klog/v2@v2.40.1: Get "https://proxy.golang.org/k8s.io/klog/v2/@v/v2.40.1.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:55080->[::1]:53: read: connection refused
       For full logs, run 'nix log /nix/store/h0bnm7dvj7jg6vhg86a9lalyqnz687wb-kubeseal-0.17.5.drv'.
error: 1 dependencies of derivation '/nix/store/zw75adh8k7kdqr3c938769mj55576sib-shell.drv' failed to build
```

Ha! Must be a firewall issue. Let's open port 53.

```bash
$ sudo ufw allow 53/udp
$ sudo ufw allow 53/tcp
```

But if you run `nix build` again...you should get the same error as before. Ouch!


### Surprise

Let's do some debugging. We're going to get an environment similar to what Nix uses to build the package and then run the Nix standard build in there---i.e. pretty much what `mkDerivation` does.

```bash
$ nix develop .#kubeseal
$ source $stdenv/setup
$ genericBuild
```

This build actually works

```bash
$ ./kubeseal --version
kubeseal version: 0.17.5
$ exit
```

What the heck?! Let's try again but with a cleaner environment.

```bash
$ sudo rm -rf outputs source
$ nix develop -i .#kubeseal
$ source $stdenv/setup
$ genericBuild
unpacking sources
unpacking source archive /nix/store/ilv8ribp7ic29hql6lszpp35qyxjn429-source
source root is source
patching sources
configuring
no configure script, doing nothing
building
missing $GOPATH
```

Fair enough, let's set `GOPATH`.

```bash
$ sudo rm -rf outputs source
$ mkdir ignore-me
$ nix develop -i .#kubeseal
$ export GOPATH=`pwd`/ignore-me
$ source $stdenv/setup
$ genericBuild
```

This actually works

```bash
$ ./kubeseal --version
kubeseal version: 0.17.5
$ exit
```

But why?! Hang on though. It gets better. Given that setting `GOPATH` worked in what's supposedly an environment as close as it gets to the actual Nix build, I thought adding these two lines to the build phase in `kubeseal.nix` would to the trick

```nix
...
  buildPhase = ''
    mkdir ignore-me
    export GOPATH=`pwd`/ignore-me
...
```

Turns out I was so wrong. Running `nix build` again with the above change outputs the same "connection refused" error as before. Back to square one.

--- a/nix/flake.nix
+++ b/nix/flake.nix
@@ -3,19 +3,22 @@
 
   inputs = {
     nixpkgs.url = "github:NixOs/nixpkgs/nixos-21.11";
+    nixpkgs-22-05.url = "github:NixOs/nixpkgs/nixos-22.05";
     nixie = {
       url = "github:c0c0n3/nixie";
       inputs.nixpkgs.follows = "nixpkgs";
     };
   };
 
-  outputs = { self, nixpkgs, nixie }:
+  outputs = { self, nixpkgs, nixpkgs-22-05, nixie }:
     let
       buildWith = nixie.lib.flakes.mkOutputSetForCoreSystems nixpkgs;
       mkSysOutput = { system, sysPkgs }:
       let
         opa = sysPkgs.callPackage ./opa.nix {};
-        kubeseal = sysPkgs.callPackage ./kubeseal.nix {};
+        # kubeseal = sysPkgs.callPackage ./kubeseal.nix {};  # NOTE (1)
+        kubeseal = nixpkgs-22-05.legacyPackages.${system}.kubeseal;
+        # ^ Nixpkgs 22.05 comes w/ kubeseal 0.17.5, same as kubeseal.nix
       in {
         defaultPackage.${system} = with sysPkgs; buildEnv {
           name = "kitt4sme-cluster-shell";
@@ -25,3 +28,10 @@
     in
       buildWith mkSysOutput;
 }
+# NOTE
+# 1. kubeseal. Ideally we'd build it off Nixpkgs 21.11, our main package
+# source at the moment. In fact, that'd save quite a bit of build time
+# and almost 4GB space in the Nix store---on MacOS. That's why we whipped
+# together that kubeseal.nix recipe. But it looks like we did it too quickly,
+# as it breaks on some Linux flavours, see
+# - https://github.com/c0c0n3/kitt4sme.live/issues/167

